Sources/VSM/StateContainer+Debug.swift

```diff
@@ -49,9 +37,105 @@ private enum _StateContainerDebugLogger {
 extension StateContainer {
     func registerForDebugLogging() {
 #if DEBUG
-        if _StateContainerDebugLogger.isLoggingAllStateContainers {
-            _StateContainerDebugLogger.register(stateContainer: self)
+        if !StateContainerDebugLogger.defaultLoggingModes.isEmpty {
+            debugLogger.startLogging(for: self, options: StateContainerDebugLogger.defaultLoggingModes)
         }
 #endif
     }
 }
+
+/// Manages debug logging for a state container
+class StateContainerDebugLogger {
```

Comment:
Perhaps for a different time, but long term I think Apple's [universal logging system](https://developer.apple.com/documentation/os/logging) would be beneficial here. In short, you can use [os_log](https://developer.apple.com/documentation/os/2320718-os_log) to send messages to various system tools including the Xcode console, Console app, and Instruments. It's designed to be highly performant so that it can be left in production code.  So we could embed at least some of these operations into `StateContainer` itself instead of using publishers or having to enable it per container.  (You can still enable/disable the logging based on the `OSLog` used).  This is especially useful when profiling a release build of the app in Instruments. Of course, if the operations surrounding the logs aren't performant we would still want to have some guardrails.

Combine this with os_signposts and a custom instrument for using in Instruments and we could have a whole profiling toolkit.

Reply:
This would be an excellent addition! My only concern would be that of the potential of having noisy logs by default. In some features, state updates are _very_ frequently, which would make it difficult to see useful information from the Xcode console.